lms/djangoapps/courseware/access.py

```diff
@@ -250,20 +248,30 @@ def _can_enroll_courselike(user, courselike):
     # which actually points to a CourseKey. Sigh.
     course_key = courselike.id
 
+    course_enrollment_open = courselike.is_enrollment_open()
```

Comment:
@Agrendalath 
Do you think we should implement it like this, to maintain loose coupling ?
```suggestion
    course_enrollment_open = course_metadata_utils.is_enrollment_open(courselike.enrollment_start, courselike.enrollment_end)
```

Reply:
@kaustavb12, what would be the advantage of the suggested approach? The current one seems good to me.

Reply:
@Agrendalath 

The only `courselike` objects currently being passed here seem to be `CourseBlock` and `CourseOverview` and I have added the `is_enrollment_open()` method to both.

However, it seems that the [_can_enroll_courselike](https://github.com/open-craft/edx-platform/blob/a76fa298011610313da7c592d08a0981c72ef5cc/lms/djangoapps/courseware/access.py#L235-L291) function is intended to be loosely coupled, with the caller able to pass any object as `courselike` with the bare minimum attributes. The current change adds one more such attribute. ie. `is_enrollment_open` which `courselike` objects must have.

My concern here arose due to [this test case](https://github.com/openedx/edx-platform/blob/068eecff91cbdfc2fe6c6770a7acb3985e7fa654/lms/djangoapps/courseware/tests/test_access.py#L462-L466), where a mock was passed enrollment start and end dates. However, with the current changes, I had to [use the CourseOverviewFactory](https://github.com/open-craft/edx-platform/blob/a76fa298011610313da7c592d08a0981c72ef5cc/lms/djangoapps/courseware/tests/test_access.py#L464-L469) instead in the test case.

Reply:
@kaustavb12, the docstring defines [which classes are accepted as a courselike](https://github.com/openedx/edx-platform/blob/a76fa298011610313da7c592d08a0981c72ef5cc/lms/djangoapps/courseware/access.py#L241), so I wouldn't worry about this too much.